File: crates/sc-search/src/codec.rs
//! Block compression (§4.1).
//!
//! zstd, via the `zstd` crate. `zstd::bulk` is the right API here because every
//! payload is a self-contained, known-length block: no streaming state, one
//! call in each direction.
//!
//! The C binding is not a problem for this crate — `DESIGN-PREVIEW.md`'s
//! "pure Rust decoders only" rule is about parsers that eat untrusted user
//! content (images, documents). These bytes are ours: we wrote them, in a file
//! we own, and every decompression is length-checked against the block
//! directory before it is trusted.

use anyhow::{bail, Context, Result};

/// Base-segment blocks. Read once per candidate hit and written once per
/// build/merge, so ratio matters more than encode speed. Level 6 measured
/// meaningfully better than 3 on filename corpora at a cost paid only at build
/// time.
pub const BASE_LEVEL: i32 = 6;

/// Delta records. §4.2 asks for "light compression or none" — delta segments are
/// linearly scanned on *every* query, so decode cost dominates.
pub const DELTA_LEVEL: i32 = 1;

/// Ceiling on a single decompressed block, so a corrupt length prefix cannot
/// make us allocate the machine. A 32-name block of 4 KiB paths is ~128 KiB;
/// 64 MiB is four orders of magnitude of headroom.
pub const MAX_DECOMPRESSED: usize = 64 << 20;

pub fn compress(data: &[u8]) -> Vec<u8> {
    zstd::bulk::compress(data, BASE_LEVEL).expect("zstd compression of an in-memory buffer")
}

pub fn compress_fast(data: &[u8]) -> Vec<u8> {
    zstd::bulk::compress(data, DELTA_LEVEL).expect("zstd compression of an in-memory buffer")
}

/// Decompress a payload whose uncompressed length is not known up front.
pub fn decompress(data: &[u8]) -> Result<Vec<u8>> {
    let out = zstd::bulk::decompress(data, MAX_DECOMPRESSED)
        .context("zstd stream is truncated or corrupt")?;
    Ok(out)
}

/// Decompress with the uncompressed length from the block directory. The
/// length is both an allocation hint and a check: a block that does not
/// decompress to exactly the recorded size is not the block we wrote.
pub fn decompress_hint(data: &[u8], expect: usize) -> Result<Vec<u8>> {
    if expect > MAX_DECOMPRESSED {
        bail!("declared block size {expect} exceeds {MAX_DECOMPRESSED}");
    }
    let cap = if expect == 0 { MAX_DECOMPRESSED } else { expect };
    let out = zstd::bulk::decompress(data, cap).context("zstd stream is truncated or corrupt")?;
    if expect != 0 && out.len() != expect {
        bail!(
            "block directory says {expect} bytes, stream produced {}",
            out.len()
        );
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip() {
        let data = b"IMG_0001.jpg\nIMG_0002.jpg\nIMG_0003.jpg\n".repeat(40);
        let c = compress(&data);
        assert!(
            c.len() < data.len(),
            "no compression win: {} vs {}",
            c.len(),
            data.len()
        );
        assert_eq!(decompress_hint(&c, data.len()).unwrap(), data);
    }

    #[test]
    fn delta_level_roundtrips_too() {
        let data = b"docs/report-2026.pdf".repeat(20);
        assert_eq!(decompress(&compress_fast(&data)).unwrap(), data);
    }

    #[test]
    fn tree_order_adjacency_is_where_the_win_is() {
        // §4.1 point 2: the compression win comes from prefix sharing between
        // adjacent names, which only exists if blocks are built in tree order.
        let mut ordered = Vec::new();
        for i in 0..32 {
            ordered.extend_from_slice(format!("photos/2026/summer/IMG_{i:04}.jpg\n").as_bytes());
        }
        let mut scattered = Vec::new();
        for i in 0..32u64 {
            let h = crate::hll::hash64(&i.to_le_bytes());
            scattered.extend_from_slice(format!("{h:016x}/{:016x}.dat\n", h.rotate_left(21)).as_bytes());
        }
        let a = compress(&ordered).len() as f64 / ordered.len() as f64;
        let b = compress(&scattered).len() as f64 / scattered.len() as f64;
        assert!(a < b, "tree-ordered ratio {a:.3} should beat scattered {b:.3}");
    }

    #[test]
    fn corrupt_input_errors_rather_than_panics() {
        assert!(decompress(b"not a zstd frame at all").is_err());
    }

    #[test]
    fn length_mismatch_is_rejected() {
        let c = compress(b"hello");
        assert!(decompress_hint(&c, 999).is_err());
    }
}

File: go/internal/search/index/codec.go
// Package index implements the trigram index: the immutable base segment, the
// delta and tombstone overlay, and the union that answers a query.
//
// The on-disk format is fixed. Every structural byte this package writes is
// compared against a fixture the Rust implementation produced, so a change
// here is a change to a format two implementations have to agree on.
package index

import (
	"errors"
	"fmt"
	"sync"

	"github.com/klauspost/compress/zstd"
)

// Block compression.
//
// zstd's format specifies what a decoder accepts rather than what an encoder
// emits, so two independent encoders are not required to agree byte for byte.
// What is required, and what the fixtures check, is that each side decodes the
// other's frames to identical bytes of the recorded length.

// BaseLevel compresses base-segment blocks. They are read once per candidate
// hit and written once per build, so ratio matters more than encode speed.
const BaseLevel = zstd.SpeedBetterCompression

// DeltaLevel compresses delta records, which are linearly scanned on every
// query, so decode cost dominates and the level is the fastest one.
const DeltaLevel = zstd.SpeedFastest

// MaxDecompressed is the ceiling on one decompressed block, so a corrupt
// length prefix cannot make this allocate the machine. A 32-name block of 4 KiB
// paths is about 128 KiB, so this is four orders of magnitude of headroom.
const MaxDecompressed = 64 << 20

// ErrCorrupt is a payload that is not what the directory says it is.
var ErrCorrupt = errors.New("index: a corrupt segment")

// The encoders and decoders are stateless across calls and safe for concurrent
// use, and building one costs an allocation of window-sized buffers, so they
// are made once.
//
//nolint:gochecknoglobals // zstd encoders and decoders are documented safe for concurrent use.
var (
	baseEncoder  = sync.OnceValues(func() (*zstd.Encoder, error) { return newEncoder(BaseLevel) })
	deltaEncoder = sync.OnceValues(func() (*zstd.Encoder, error) { return newEncoder(DeltaLevel) })
	sharedDecode = sync.OnceValues(func() (*zstd.Decoder, error) {
		return zstd.NewReader(nil,
			zstd.WithDecoderConcurrency(0),
			zstd.WithDecoderMaxMemory(MaxDecompressed))
	})
)

func newEncoder(level zstd.EncoderLevel) (*zstd.Encoder, error) {
	return zstd.NewWriter(nil, zstd.WithEncoderLevel(level), zstd.WithEncoderConcurrency(1))
}

// Compress encodes a base-segment block.
func Compress(data []byte) ([]byte, error) {
	e, err := baseEncoder()
	if err != nil {
		return nil, fmt.Errorf("index: building the encoder: %w", err)
	}
	return e.EncodeAll(data, nil), nil
}

// CompressFast encodes a delta record.
func CompressFast(data []byte) ([]byte, error) {
	e, err := deltaEncoder()
	if err != nil {
		return nil, fmt.Errorf("index: building the encoder: %w", err)
	}
	return e.EncodeAll(data, nil), nil
}

// Decompress decodes a payload whose uncompressed length is not known.
func Decompress(data []byte) ([]byte, error) {
	d, err := sharedDecode()
	if err != nil {
		return nil, fmt.Errorf("index: building the decoder: %w", err)
	}
	out, err := d.DecodeAll(data, nil)
	if err != nil {
		return nil, fmt.Errorf("%w: the stream is truncated or corrupt: %w", ErrCorrupt, err)
	}
	return out, nil
}

// DecompressHint decodes a payload whose uncompressed length the block
// directory recorded.
//
// The length is both an allocation hint and a check: a block that does not
// decompress to exactly the recorded size is not the block that was written,
// and continuing with it would mean parsing names out of something else.
func DecompressHint(data []byte, expect int) ([]byte, error) {
	if expect < 0 || expect > MaxDecompressed {
		return nil, fmt.Errorf("%w: a declared block size of %d", ErrCorrupt, expect)
	}
	d, err := sharedDecode()
	if err != nil {
		return nil, fmt.Errorf("index: building the decoder: %w", err)
	}
	var dst []byte
	if expect > 0 {
		dst = make([]byte, 0, expect)
	}
	out, err := d.DecodeAll(data, dst)
	if err != nil {
		return nil, fmt.Errorf("%w: the stream is truncated or corrupt: %w", ErrCorrupt, err)
	}
	if expect != 0 && len(out) != expect {
		return nil, fmt.Errorf("%w: the directory says %d bytes and the stream produced %d",
			ErrCorrupt, expect, len(out))
	}
	return out, nil
}
